ledger/object/sequence.go

```diff
@@ -0,0 +1,391 @@
+}
+
+func (pos SequenceRecordPosition) GreaterThan(other SequenceRecordPosition) bool {
+	return pos.PulseNumber > other.PulseNumber ||
+		(pos.PulseNumber == other.PulseNumber && pos.Index > other.Index)
+}
+
+type SequenceRecord struct {
+	Position SequenceRecordPosition
+	ID       insolar.ID
+	Record   record.MaterialRecord
+}
+
+type SequenceRecordAccessor interface {
+	// ForPosition returns sequence record for provided record position.
+	ForPosition(ctx context.Context, pos SequenceRecordPosition) (SequenceRecord, error)
+	// FromPosition returns sequence record stream from provided record position with corresponding limit.
+	FromPosition(ctx context.Context, pos SequenceRecordPosition, limit uint32) chan interface{}
+}
+
+type SequenceRecordModifier interface {
+	// Push saves new record-value in storage and increment topIndex position.
+	Push(ctx context.Context, pn insolar.PulseNumber, id insolar.ID) error
+}
+
+type SequenceRecordCursor interface {
+	// Top returns the greatest position for sequence records.
+	Top(ctx context.Context) (SequenceRecordPosition, error)
+	// TopSync returns position for sequence records at last synced pulse.
+	TopSync(ctx context.Context) (SequenceRecordPosition, error)
+}
+
+type SequenceRecordStorage interface {
+	SequenceRecordAccessor
+	SequenceRecordModifier
+	SequenceRecordCursor
+
+	// UpdatePulse makes "upsert" for top pulse number and sets at old pulse descriptor link to next pulse
+	UpdatePulse(ctx context.Context, pn insolar.PulseNumber) error
+}
+
+type SequenceRecordMemory struct{}
+
+type SequenceRecordDB struct {
+	lock    sync.RWMutex
+	db      store.DB
+	top     SequenceRecordPosition
+	records RecordAccessor
+}
+
+// Push saves new record-value in storage and increment topIndex position.
+func (s *SequenceRecordDB) Push(ctx context.Context, pn insolar.PulseNumber, id insolar.ID) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	pd, err := s.exactPulse(pn)
+	if err == nil {
+		pd.topIndex++
+	} else {
+		pd = pulseLabel{curr: pn, next: 0, topIndex: 0}
+	}
+	err = s.updatePulseDescriptor(pn, pd)
+	if err != nil {
+		return errors.Wrap(err, "[SequenceRecordIndex] failed to update pulse descriptor")
+	}
+	pos := SequenceRecordPosition{Index: pd.topIndex, PulseNumber: pn}
+	err = s.set(pos, id)
+	if err != nil {
+		return errors.Wrapf(err, "[SequenceRecordIndex] failed to save record at position: %s", pos)
+	}
+	if pos.GreaterThan(s.top) {
+		s.top = pos
+	}
+	return nil
+}
+
+// ForPosition returns sequence record for provided record position.
+func (s *SequenceRecordDB) ForPosition(ctx context.Context, pos SequenceRecordPosition) (SequenceRecord, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	return s.get(ctx, pos)
+}
+
+// FromPosition returns sequence record stream from provided record position with corresponding limit.
+func (s *SequenceRecordDB) FromPosition(ctx context.Context, pos SequenceRecordPosition, limit uint32) chan interface{} {
```

Comment:
I think, It's a good idea to return empty chan of interfaces with two different values.
It'd be better to return two channels, with exact result and error

Reply:
Okay, but I wanted to do something similar to how it works in java Reactive Streams (projectreactor.io) when data and errors occurring during the processing of a stream are transmitted through a single stream. What are the pitfalls here?

Reply:
It's not  java. Working through empty interfaces isn't a great thing in go. https://go-proverbs.github.io/

In our experience, casting of empty interfaces is an infinite root of problems

Reply:
Using empty interfaces is generally frowned upon in Go community. Let's consider other options.

Reply:
Okay, I replaced it with a simple slice and error value